### PR TITLE
feat: release stabilization

### PR DESCRIPTION
**Pull request by p0deje**
Bundles multiple fixes that are required to successfully release Python and TypeScript packages with a new `alumnium` binary.

### CODE REVIEW

File: .github/workflows/ci-typescript.yml [hunk 1]
```diff
@@ -50,10 +50,6 @@ jobs:
           - driver: appium-ios
             model: azure_openai
 
-          # TODO: Add support for appium-android to system tests
-          # - driver: appium-android
-          #   model: azure_openai
-
```
Comment (kossnocorp) [inline .github/workflows/ci-typescript.yml:56]: It would be great to have it or it's plain impossible?

File: packages/typescript/scripts/build-pip-src.sh [hunk 2]
```diff
@@ -10,24 +10,31 @@ fi
...
 DIST_DIR="$PKG_DIR/dist"
-SRC_TAR_GZ_NAME="alumnium-cli-$VERSION.tar.gz"
+SRC_TAR_GZ_NAME="alumnium_cli-$VERSION.tar.gz"
 SRC_TAR_GZ_PATH="$DIST_DIR/pip/$SRC_TAR_GZ_NAME"
 
 tmp_dir="$(mktemp -d)"
 trap 'rm -rf "$tmp_dir"' EXIT
 
+# PyPI sdist requires a single top-level directory named {package}-{version}
+top_dir="$tmp_dir/alumnium_cli-$VERSION"
```
Comment (kossnocorp) [inline packages/typescript/scripts/build-pip-src.sh:20]: What's top btw?

File: packages/typescript/scripts/build.ts [hunk 3]
```diff
@@ -186,49 +200,71 @@ const loggerPathPlugin: BunPlugin = {
...
+      async (args) => {
+        return {
+          contents: await Bun.file(args.path).bytes(),
+          loader: "file",
+        };
+      },
+    );
+  },
+};
+
```
Comment (kossnocorp) [inline packages/typescript/scripts/build.ts:267]: Out of scope, but we need to add tests that check these edge cases. They are very obscure!

File: packages/typescript/scripts/publish.sh [hunk 4]
```diff
@@ -24,7 +24,7 @@ publish_pkg() {
 	[[ -f "$pkg" ]] || return 0
 
 	echo -e "🌀️ Publishing $pkg...\n"
-	PUBLISH_ARGS=("--provenance" "--access public")
+	PUBLISH_ARGS=("--provenance", "--access=public")
```
Comment (kossnocorp) [inline packages/typescript/scripts/publish.sh]: It adds `,` to `--provenance`. Bash arrays have no commas.

```suggestion
	PUBLISH_ARGS=("--provenance" "--access=public")
```

File: packages/typescript/scripts/build.ts [hunk 5]
```diff
@@ -355,9 +395,7 @@ async function main() {
         distPackageJson.bin = distPackageJson.publishConfig.bin;
 
         TARGET_PLATFORMS.forEach(({ npm }) => {
-          distPackageJson.optionalDependencies[npm.name] = `file:${npm.dir}`;
-          distPackageJson.publishConfig.optionalDependencies[npm.name] =
-            ALUMNIUM_VERSION;
+          distPackageJson.optionalDependencies[npm.name] = ALUMNIUM_VERSION;
```
Comment (kossnocorp) [inline packages/typescript/scripts/build.ts:398]: I feel like removing `publishConfig` forced you to add these lines to the npm tests below, which makes the test more explicit. Doesn't Bun use `publishConfig` when packing?

```
  "pnpm": {
    "overrides": {
      "@alumnium/cli-darwin-arm64": "file:../../../dist/npm-alumnium-cli-darwin-arm64",
      "@alumnium/cli-darwin-x64": "file:../../../dist/npm-alumnium-cli-darwin-x64",
      "@alumnium/cli-linux-arm64": "file:../../../dist/npm-alumnium-cli-linux-arm64",
      "@alumnium/cli-linux-x64": "file:../../../dist/npm-alumnium-cli-linux-x64",
      "@alumnium/cli-windows-arm64": "file:../../../dist/npm-alumnium-cli-windows-arm64",
      "@alumnium/cli-windows-x64": "file:../../../dist/npm-alumnium-cli-windows-x64"
    }
```

File: packages/typescript/scripts/build.ts [hunk 6]
```diff
@@ -748,4 +809,58 @@ function getNpmOs(os: OS) {
...
+> {
+  const seleniumPkgDir = await fs.realpath(
+    path.resolve(PKG_DIR, "node_modules/selenium-webdriver"),
+  );
+  const playwrightCorePkgDir = await fs.realpath(
+    path.resolve(PKG_DIR, "node_modules/playwright-core"),
+  );
+
+  const seleniumAtomPaths = Array.from(
+    new Bun.Glob(path.join(seleniumPkgDir, "lib/atoms/*.js")).scanSync("/"),
```
Comment (kossnocorp) [inline packages/typescript/scripts/build.ts]: It's better to use async even if in practice nothing can get blocked.

File: packages/typescript/scripts/build.ts [hunk 7]
```diff
@@ -748,4 +809,58 @@ function getNpmOs(os: OS) {
...
+
+async function getStandaloneEmbeddedAssets(): Promise<
+  StandaloneEmbeddedAsset[]
+> {
+  const seleniumPkgDir = await fs.realpath(
+    path.resolve(PKG_DIR, "node_modules/selenium-webdriver"),
+  );
+  const playwrightCorePkgDir = await fs.realpath(
+    path.resolve(PKG_DIR, "node_modules/playwright-core"),
+  );
```
Comment (kossnocorp) [inline packages/typescript/scripts/build.ts:836]: Do you have to use `fs.realpath` because of symlinks? Worth adding a note.

File: packages/typescript/scripts/build.ts [hunk 8]
```diff
@@ -266,19 +302,23 @@ async function main() {
...
     console.log("\n🌀 Building binaries:\n");
 
+    const standaloneEmbeddedAssetPaths =
+      await prepareStandaloneEmbeddedAssets();
+
     await Promise.all(
       TARGET_PLATFORMS.map(async ({ os, arch, target, binPath }) => {
         const result = await Bun.build({
-          entrypoints: [BIN_SRC_PATH],
+          entrypoints: [BIN_SRC_PATH, ...standaloneEmbeddedAssetPaths],
```
Comment (kossnocorp) [inline packages/typescript/scripts/build.ts:311]: How does adding `standaloneEmbeddedAssetPaths` to `entrypoints` work? Worth adding a note. From the code it seems like you're pointing Bun to compile/embed Selenium binaries, but I checked bin sizes and they haven't changed a bit.

File: packages/typescript/scripts/build.ts [hunk 9]
```diff
@@ -620,7 +658,30 @@ function getPyProjectToml(project: PyProject) {
...
+    "Summary: Alumnium CLI",
+    `Author: ${authorNames}`,
+    `Author-email: ${authorEmails}`,
+    "License-Expression: MIT",
+    "Requires-Python: >=3.10, <4.0",
+    "Project-URL: Documentation, https://alumnium.ai/docs/",
+    "Project-URL: Homepage, https://alumnium.ai/",
+    "Project-URL: Issues, https://github.com/alumnium-hq/alumnium/issues",
+    "Project-URL: Repository, https://github.com/alumnium-hq/alumnium",
+  ].join("\n");
```
Comment (kossnocorp) [inline packages/typescript/scripts/build.ts:684]: I would use string literal instead to get rid of the noise:

```suggestion
  return `Metadata-Version: 2.4
Name: ${PIP_CLI_PKG_NAME}
Version: ${version}
Summary: Alumnium CLI
Author: ${authorNames}
Author-email: ${authorEmails}
License-Expression: MIT
Requires-Python: >=3.10, <4.0
Project-URL: Documentation, https://alumnium.ai/docs/
Project-URL: Homepage, https://alumnium.ai/
Project-URL: Issues, https://github.com/alumnium-hq/alumnium/issues
Project-URL: Repository, https://github.com/alumnium-hq/alumnium
`;
```

File: packages/typescript/src/standalone/setupEmbeddedDependencies.ts [hunk 10]
```diff
@@ -0,0 +1,217 @@
...
+          name.slice(SELENIUM_ATOM_ASSET_PREFIX.length),
+        ),
+      ),
+    ),
+  );
+
+  const seleniumManagerPath = await extractSeleniumManager(
+    filesByName,
+    extractedDir,
+  );
```
Comment (kossnocorp) [inline packages/typescript/src/standalone/setupEmbeddedDependencies.ts]: Can't we put it all in a single `Promise.all`?

```suggestion
  const filesByName = getEmbeddedFilesByName();

  const seleniumAtomNames = Object.keys(filesByName)
    .filter((name) => name.startsWith(SELENIUM_ATOM_ASSET_PREFIX))
    .sort();

  const [seleniumManagerPath] = await Promise.all([
    extractSeleniumManager(filesByName, extractedDir),

    ...seleniumAtomNames.map((name) =>
      writeEmbeddedFile(
        filesByName,
        name,
        path.join(
          seleniumAtomsDir,
          name.slice(SELENIUM_ATOM_ASSET_PREFIX.length),
        ),
      ),
    ),

    writeEmbeddedFile(
      filesByName,
      PLAYWRIGHT_CORE_PACKAGE_JSON_ASSET_NAME,
      playwrightPackageJsonPath,
    ),
  ]);
```

File: packages/typescript/src/standalone/setupEmbeddedDependencies.ts [hunk 11]
```diff
@@ -0,0 +1,217 @@
...
+    _resolveFilename(...args: unknown[]): string;
+  };
+
+  const resolveFilename = moduleWithResolveFilename._resolveFilename;
+
+  moduleWithResolveFilename._resolveFilename = function (...args: unknown[]) {
+    const [request] = args;
+
+    if (typeof request === "string") {
+      if (request.startsWith("./atoms/")) {
```
Comment (kossnocorp) [inline packages/typescript/src/standalone/setupEmbeddedDependencies.ts:154]: I see that you check `"\\playwright-core\\package.json"` below. I assume that's for Windows. Should it use `path.sep` here too?

File: packages/typescript/src/standalone/setupEmbeddedDependencies.ts [hunk 12]
```diff
@@ -0,0 +1,217 @@
...
+        return path.join(
+          paths.seleniumAtomsDir,
+          request.slice("./atoms/".length),
```
Comment (kossnocorp) [inline packages/typescript/src/standalone/setupEmbeddedDependencies.ts:157]: Same here.

File: packages/typescript/src/standalone/setupEmbeddedDependencies.ts [hunk 13]
```diff
@@ -0,0 +1,217 @@
...
+        );
+      }
+
+      if (request === "../../../package.json") {
```
Comment (kossnocorp) [inline packages/typescript/src/standalone/setupEmbeddedDependencies.ts]: And here.

File: packages/typescript/src/standalone/setupEmbeddedDependencies.ts [hunk 14]
```diff
@@ -0,0 +1,217 @@
...
+        return paths.playwrightPackageJsonPath;
+      }
+
+      if (
+        request.endsWith("/playwright-core/package.json") ||
+        request.endsWith("\\playwright-core\\package.json")
```
Comment (kossnocorp) [inline packages/typescript/src/standalone/setupEmbeddedDependencies.ts]: It would be easier to use `join` here `path.sep + path.join("playwright-core", "package.json"`. I feel like the leading `path.sep` can be trimmed though.

File: packages/typescript/src/standalone/setupEmbeddedDependencies.ts [hunk 15]
```diff
@@ -0,0 +1,217 @@
...
+    throw new Error(`Missing embedded dependency asset: ${embeddedFileName}`);
+  }
+
+  await fs.mkdir(path.dirname(targetPath), { recursive: true });
+  await fs.writeFile(targetPath, await file.bytes());
+
+  if (mode) {
+    await fs.chmod(targetPath, mode);
+  }
+}
```
Comment (kossnocorp) [inline packages/typescript/src/standalone/setupEmbeddedDependencies.ts:206]: That's some proper dark magic, respect 🫡

**Review by kossnocorp — CHANGES_REQUESTED** [review]
The only critical-looking things are:

1. The OS-specific separator usage inside `setupEmbeddedDependencies.ts`, but it's worth checking if Bun doesn't normalize them into URLs.
2. Extra comma in a Bash array.

Everything else is optional.

File: .github/workflows/ci-typescript.yml (hunk 1 above)
Comment (p0deje) [reply .github/workflows/ci-typescript.yml:56]: It's possible, but either I need to move native app tests from Python to TypeScript, or figure out how to make Chrome browser work on Android.

File: packages/typescript/scripts/publish.sh (hunk 4 above)
Comment (p0deje) [reply packages/typescript/scripts/publish.sh]: Good catch!

File: packages/typescript/scripts/build.ts (hunk 5 above)
Comment (p0deje) [reply packages/typescript/scripts/build.ts:398]: The problem was that `publishConfig.optionalDependencies` somehow was completely ignored. When I pushed the built package to NPM and tried using it, I was getting errors because it tried installing from `file:`

```
$ npx alumnium@0.20.0-alpha.3
node:internal/modules/package_json_reader:316
  throw new ERR_MODULE_NOT_FOUND(packageName, fileURLToPath(base), null);
        ^

Error [ERR_MODULE_NOT_FOUND]: Cannot find package '@alumnium/cli-darwin-arm64' imported from /Users/p0deje/node_modules/alumnium/src/cli/binWrapper.js
    at Object.getPackageJSONURL (node:internal/modules/package_json_reader:316:9)
    at packageResolve (node:internal/modules/esm/resolve:768:81)
    at moduleResolve (node:internal/modules/esm/resolve:859:18)
    at defaultResolve (node:internal/modules/esm/resolve:991:11)
    at #cachedDefaultResolve (node:internal/modules/esm/loader:712:20)
    at #resolveAndMaybeBlockOnLoaderThread (node:internal/modules/esm/loader:729:38)
    at ModuleLoader.resolveSync (node:internal/modules/esm/loader:758:52)
    at #resolve (node:internal/modules/esm/loader:694:17)
    at ModuleLoader.getOrCreateModuleJob (node:internal/modules/esm/loader:614:35)
    at onImport.tracePromise.__proto__ (node:internal/modules/esm/loader:643:32) {
  code: 'ERR_MODULE_NOT_FOUND'
}
```

File: packages/typescript/scripts/build.ts (hunk 8 above)
Comment (p0deje) [reply packages/typescript/scripts/build.ts:311]: I've found that if you add static files, they are considered "imported" and will be added to the binary and Bun will know how to resolve them later in runtime when they are actually imported. https://bun.sh/blog/bun-v1.1.25#embedding-files-in-standalone-executables-without-importing

I feel like theses files were always there, but Bun could explain how to resolve path to them in runtime.

File: packages/typescript/scripts/build.ts (hunk 9 above)
Comment (p0deje) [reply packages/typescript/scripts/build.ts:684]: I'll keep array, don't want to mess with indentation.

File: packages/typescript/src/standalone/setupEmbeddedDependencies.ts (hunk 11 above)
Comment (p0deje) [reply packages/typescript/src/standalone/setupEmbeddedDependencies.ts:154]: I actually didn't test Windows support at all, let me revert it now and then test+fix separately.

File: packages/typescript/src/standalone/setupEmbeddedDependencies.ts (hunk 15 above)
Comment (p0deje) [reply packages/typescript/src/standalone/setupEmbeddedDependencies.ts:206]: I wish Bun had some sort of virtual filesystem so that we don't have to monkey patch! https://github.com/oven-sh/bun/issues/5445

File: packages/typescript/scripts/build-pip-src.sh (hunk 2 above)
Comment (p0deje) [reply packages/typescript/scripts/build-pip-src.sh:20]: Screenshot 2026-04-14 at 07 38 42

File: packages/typescript/scripts/build.ts (hunk 3 above)
Comment (p0deje) [reply packages/typescript/scripts/build.ts:267]: I agree, though it's very hard to test those w/o publishing the package!

File: packages/typescript/src/standalone/setupEmbeddedDependencies.ts (hunk 15 above)
Comment (kossnocorp) [reply packages/typescript/src/standalone/setupEmbeddedDependencies.ts:206]: It actually DOES have it https://bun.com/docs/bundler#mix-disk-and-virtual-files
